sklearn_pandas/categorical_imputer.py

```diff
@@ -33,16 +33,23 @@ class CategoricalImputer(BaseEstimator, TransformerMixin):
     copy : boolean, optional (default=True)
         If True, a copy of X will be created.
 
+    replacement : string, optional (default="mode")
+        The value that all occurrences of missing values are replaced
+        with. Use this to specify an alternative if you don't want to
+        impute with the mode, or if there are multiple modes in your
+        data and you want to choose a particular one.
+
     Attributes
     ----------
     fill_ : str
         Most frequent value of the training data.
 
     """
 
-    def __init__(self, missing_values='NaN', copy=True):
+    def __init__(self, missing_values='NaN', replacement='mode', copy=True):
```

Comment:
There is only one problem with this: it doesn't allow replacing all values with the string "mode". ;)  Sounds a bit stupid but it's important for it to be clean. :)

My proposal is to add two attributes:
- `strategy`: with "most_frequent" and "fixed_value" as possible options. Default must be "most_frequent", as in `sklearn.preprocessing.Imputer`.
- `replacement`: This is required when `strategy=fixed_value`. If it is not provided, an exception like `ValueError` should be raised in `__init__`.

Examples:
```
imputer_1 = CategoricalImputer(strategy="most_frequent")
imputer_2 = CategoricalImputer(strategy="fixed_value", replacement="my_value")
broken_imputer = CategoricalImputer(strategy="fixed_value")  # raises ValueError
```

You can test that a certain exception is raised in a test like:
```
with pytest.raises(ValueError):
    CategoricalImputer(strategy="fixed_value")
```

Reply:
I thought of that but I wasn't sure whether it was worth adding all that extra logic for that one corner case. I'll see what I can do.

Reply:
I think in order to avoid breaking backwards compatibility it is necessary to set a default value for `replacement` regardless of whether `strategy="most_frequent"` or not. That would in turn remove the need for the extra exception logic.

Reply:
Well, the fixed `replacement` should only be used if `strategy="most_frequent"`, so we don't really need to set a valid default value for `replacement`. Actually I think it's better to *not* set a valid default value for it, since I'd expect the user to always specify a `replacement` whenever (s)he chooses `strategy="most_frequent"`.

Reply:
Currently, using CategoricalImputer without specifying any parameters is valid syntax and imputes missing values with the mode. I suspect this is the way it is most commonly used.

Unless the new parameters are introduced as keyword parameters (which as far as I know always require default values), this will become invalid syntax, which seems likely to break lots of existing projects.

Is it acceptable to break backwards compatibility in this way, or do you have another proposal that avoids this problem? I don't have a great deal of experience with Python or these sorts of projects, so perhaps I'm missing something here.

Reply:
Please have a look at the new commit and let me know what you think.